### ISSUE TITLE
Quick fill combobox / autocomplete add new

### ISSUE THREAD

**Issue by epicserve** [select2/select2#4287]
I've been searching and I can't seem to find a plugin or this feature built into Select2. Basically, what I'm wanting is the ability to change the "no results found" message into a "+ Add New" button that when you click on the button it pops up a bootstrap modal with the body being an `<iframe />` to a form that when submitted adds a new object to the database and then adds the new `<option />` to your `<select />`.

Here is an animated gif that shows this UI pattern being used on Quickbooks Online. Based on the CSS classes, I think they are calling it a Quick fill combobox.

![quickbooks-online-quickfill-combobox](https://cloud.githubusercontent.com/assets/191620/14302218/34401d86-fb64-11e5-8f77-beaa9349839b.gif)

This seems like it would be a really handy UI pattern that could be reused by lots of people. So my question is:
1. Is this a feature that would be good to add directly to select2? And is this feature already on the roadmap?
2. If the answer is no to the first question, then is there an official plugin policy and if so what is it? I would consider writing a plugin, but I don't want to waste my time on it if a plugin like this already exists or if their are plans already to add this type of functionally directly to select2.
3. Any tips on the best way to replace the "no results found" message with a button that has an onclick event attached to it?

Thanks to @claar for finding the example on Quickbooks online and making the animated gif.
## Prerequisites
- [x] I have searched for similar issues in both open and closed tickets and cannot find a duplicate
- [x] The issue still exists against the latest `master` branch of Select2
- [x] This is not a usage question (Those should be directed to the [community](https://select2.github.io/community.html))
- [ ] I have attempted to find the simplest possible steos to reproduce the issue
- [ ] I have included a failing test as a pull request (Optional)

Libraries
- Select2 version: 4.0.2

**Comment by epicserve** [comment]
@kevin-brown any ideas on how to add the "add button" when no results are found? I saw [this](https://github.com/select2/select2/blob/master/src/js/select2/results.js#L392) in the code, which made me think I could do something like `$my_select2_el.on('results:message', function(params) { /* custom add button code here */ }));` however, no joy.

**Comment by epicserve** [comment]
Looks like I can change the "no results" message this way:

```javascript
$my_select2_el.select2({
    "language": {
        "noResults": function() {
            return 'Testing 1, 2, 3 ...';
        }
    }
});
```

However this just displays the button html:

```javascript
$my_select2_el.select2({
    "language": {
        "noResults": function() {
            return '<button class="btn btn-default">Add New Item</button>';
        }
    }
});
```

**Comment by epicserve** [comment]
Thanks to a friend who was able to help me override the displayMessage method. So I'm getting closer ... with the follow I was able to add a button. Now I just need to figure out how to make it so a new button isn't created with every new letter typed, once the "no results" message/button has been displayed.

```javascript
 jQuery.fn.select2.amd.require('select2/results').prototype.displayMessage = function(params) {

    var escapeMarkup = this.options.get('escapeMarkup');

    this.clear();
    this.hideLoading();

    var $message = $(
      '<li role="treeitem" aria-live="assertive"' +
      ' class="select2-results__option"></li>'
    );

    var message = this.options.get('translations').get(params.message);
    message = message(params.args);

    if (message == 'No results found') {
        $message.append('<button class="btn btn-default"><span class="glyphicon glyphicon-plus-sign"></span> Add New Item</button>');
    } else {
        $message.append(escapeMarkup(message));
    }

    $message[0].className += ' select2-results__message';

    this.$results.append($message);

};
```

**Comment by epicserve** [comment]
Got it working. I'll need to clean up my code and do some refactoring, but maybe I might have something to share in a couple days. But the following is an animated gif showing how it works with a Bootstrap modal.

![select2_add_new_button](https://cloud.githubusercontent.com/assets/191620/14757508/feadd196-08b9-11e6-9037-b2098f47377e.gif)

**Comment by goodale** [comment]
This is excellent! Looking forward to the  working code

**Comment by epicserve** [comment]
I've abstracted my code which could be used by anyone to add the functionality that is show in the animated gif above. It's not the most elegant way of doing things, but until this is an officially supported way of hooking into the "no results" message, this is a way to get it done.

Also, here is a JS Fiddle I made real quick to demo it, https://jsfiddle.net/epicserve/6000kasq/

P.S. I'm not sure if any Select2 contributors with commit access are listening to this issue, but if you are, I would be happy to make a pull request that adds in some options so there is an official way to hook into Select2 for adding this functionality. Just let me know and let me know what variable and API ideas you have.

```
/**
 * Select2QuickFillCombobox - A HTML select component that uses Select2 to add a button for adding items to a select when no options are found in the search.
 * 
 * @param $select2_el                                           The jQuery select element
 * @param {Object}  [options]
 * @param {Object}  [options.no_results_add_button_callback]    A function to call after the button as been added
 * @param {Boolean} [options.no_results_add_button_template]    The button HTML template to use
 * @constructor
 */
function Select2QuickFillCombobox($select2_el, options) {
	
    this.$select2_el = $select2_el;
    this.options = options || {};
    var defaults = {
        no_results_add_button_callback: null,
        no_results_add_button_template: '' +
            '<button type="button" class="btn btn-default">' +
            '    <span class="glyphicon glyphicon-plus-sign"></span> Add Item' +
            '</button>'
    };

    var select2_options = $.extend(defaults, this.options);
    
    if (this.$select2_el.data('select2')) {
    	this.$select2_el.select2('destroy');
    }
    this.$select2_el.select2(select2_options);
    this.override_select2_display_message_method();

}

Select2QuickFillCombobox.prototype = {

    override_select2_display_message_method: function() {

        jQuery.fn.select2.amd.require('select2/results').prototype.displayMessage = function(params) {

            var escapeMarkup = this.options.get('escapeMarkup');

            this.clear();
            this.hideLoading();

            var $message = $(
              '<li role="treeitem" aria-live="assertive"' +
              ' class="select2-results__option"></li>'
            );

            var message = this.options.get('translations').get(params.message);
            var no_results_add_button_template =  this.options.get('no_results_add_button_template');
            var no_results_add_button_callback =  this.options.get('no_results_add_button_callback');
            message = message(params.args);

            if (params.message == 'noResults' && no_results_add_button_template) {

                var $add_btn = $(no_results_add_button_template);
                $message.append($add_btn);
                if (no_results_add_button_callback) {
                    no_results_add_button_callback($add_btn);
                }

            } else {
                $message.append(escapeMarkup(message));
            }

            $message[0].className += ' select2-results__message';

            this.$results.append($message);

        };

    }

};
```

**Comment by goodale** [comment]
I am looking for something just like this for the 3.5.x version of select2.   I have something primitive working but there are some issues revolving around usability.  I am curious if your select2 v4 version has conquered these:


1.       Clicking a link in the ‘formatNoMatches’ found calls a method which launches a jQuery dialogue just fine.  However, the focus is still on the select2.   Adding an ‘open:’ handler to the dialogue to set the focus to a field in the pop-up form doesn’t work.

2.       To accommodate users which prefer keyboard-only data entry I’d like to have the option for the user to hit the ‘enter’ key or ‘tab’ key if the select2 is in a state where no results are found which launches the dialogue

3.       The dialogue needs have a reference back to the specific select2 so it can programmatically update the selected item if the pop-up was able to successfully create a new record via ajax.

4.       Needs to support multiple select2 widgets on the form, each with either a different id or data-id attribute (so the pop-up dialog needs to be able to correctly update the appropriate select2 with the new key and display label for the new record.

In any case, if anyone monitoring this thread is interested in some commission work drop me a line.

Cheers

**Comment by epicserve** [comment]
@goodale I'm not sure why but when you reply via email it adds my comment to your comment, which hurts the readability of this thread. Could  you please update your comments remove my "quoted" comment from your comments? 

Regarding your questions ...

1. I'm not sure I know what you're talking about, did you try the JS Fiddle to see if it does what you're describing?

2. Good suggestion. I didn't think about support for keyboard only usage. I wouldn't think it would be that hard to add in your own custom callback.

3. My code works fine with what you're talking about. Have a look at the JS Fiddle.

4. If you need another select with this functionality just call `new Select2QuickFillCombobox()` for the select you want to add it to.

**Comment by goodale** [comment]
Thanks for the reply and apologies for the inclusion of your comment in my reply.

Regarding #1,  the jsFiddle example calls an alert() on the Add button.   In my tests the focus issue isn’t a problem with an alert (since that is a browser based modal)  but is a problem when the button event listener creates the dialogue.  For example, when I display the dialogue it pops up but the focus is still on the select2 term field.  Do you have an example where you launch an actual dialogue and successfully put the focus into the first field in the form within the dialogue?

**Comment by mtaylor091** [comment]
@epicserve Thank you for your solution it is working great! I use QBO and have envied their autocomplete for sometime. There is one feature I would like to request from you. Can you add a way so that once the select is opened, it shows the add button instead of only showing the add button when a search doesn't match? I have tried listening to the open event and then triggering some keys like this: `$search.trigger($.Event('input', { which: 13 }));` but it isn't working how I intended.

**Comment by epicserve** [comment]
Have a look at the line `if (params.message == 'noResults' && no_results_add_button_template) {`. I think you would just change the condition so it adds the button on other messages besides 'noResults', but I'm just guessing.

**Comment by epicserve** [comment]
It sure would be nice if this functionality was just built into Select2. :)

**Comment by alexweissman** [comment]
@epicserve, consider writing this as a plugin, and posting it to the [forums](https://forums.select2.org/c/ecosystem)!
